File: src/main/java/com/ternbusty/takoyaki/util/Json.java
package com.ternbusty.takoyaki.util;

import com.ternbusty.takoyaki.util.json.JsonParser;
import com.ternbusty.takoyaki.util.json.JsonWriter;

import java.io.IOException;
import java.nio.file.Files;
import java.nio.file.Path;
import java.util.function.Function;

/**
 * Thin facade over {@link JsonParser} and {@link JsonWriter} matched to
 * the callsites in takoyaki.
 *
 * <p>The shape that flows through here is always a JSON tree —
 * {@code Map<String,Object> / List<Object> / String / Long / Double /
 * Boolean / null}. Bean-shaped types (Spec, State, KontainerConfig, ...)
 * convert to/from the tree via their static {@code fromJson(Object)} and
 * instance {@code toJson()} methods. {@link #readFile} takes the
 * {@code fromJson} mapper as a {@link Function}, replacing jackson's
 * {@code readValue(in, Class)} reflection.
 */
public final class Json {
    private Json() {}

    public static <T> T readFile(Path path, Function<Object, T> fromJson) throws IOException {
        return fromJson.apply(JsonParser.parse(Files.readAllBytes(path)));
    }

    public static void writeFile(Path path, Object jsonTree) throws IOException {
        Files.writeString(path, JsonWriter.toPretty(jsonTree));
    }

    public static String encode(Object jsonTree) {
        return JsonWriter.toPretty(jsonTree);
    }

    /** Compact (single-line, no whitespace) JSON encoding. */
    public static String encodeCompact(Object jsonTree) {
        return JsonWriter.toCompact(jsonTree);
    }

    public static <T> T decode(String s, Function<Object, T> fromJson) {
        return fromJson.apply(JsonParser.parse(s));
    }
}

File: src/main/java/com/ternbusty/takoyaki/util/Json.kt
package com.ternbusty.takoyaki.util

import kotlinx.serialization.json.Json
import kotlinx.serialization.json.JsonArray
import kotlinx.serialization.json.JsonElement
import kotlinx.serialization.json.JsonNull
import kotlinx.serialization.json.JsonObject
import kotlinx.serialization.json.JsonPrimitive
import java.io.IOException
import java.nio.file.Files
import java.nio.file.Path

object JsonCodec {

    val json = Json {
        ignoreUnknownKeys = true
        explicitNulls = false
        encodeDefaults = true
    }

    val prettyJson = Json {
        ignoreUnknownKeys = true
        explicitNulls = false
        encodeDefaults = true
        prettyPrint = true
    }

    @Throws(IOException::class)
    inline fun <reified T> loadFromFile(path: Path): T =
        json.decodeFromString(Files.readString(path))

    @Throws(IOException::class)
    inline fun <reified T> saveToFile(path: Path, value: T) {
        Files.writeString(path, prettyJson.encodeToString(value))
    }

    inline fun <reified T> encode(value: T): String =
        prettyJson.encodeToString(value)

    inline fun <reified T> encodeCompact(value: T): String =
        json.encodeToString(value)

    inline fun <reified T> decode(s: String): T =
        json.decodeFromString(s)

    fun encodePretty(element: JsonElement): String =
        prettyJson.encodeToString(JsonElement.serializer(), element)

    fun encodeCompact(element: JsonElement): String =
        json.encodeToString(JsonElement.serializer(), element)

    fun toJsonElement(value: Any?): JsonElement = when (value) {
        null -> JsonNull
        is JsonElement -> value
        is String -> JsonPrimitive(value)
        is Boolean -> JsonPrimitive(value)
        is Int -> JsonPrimitive(value)
        is Long -> JsonPrimitive(value)
        is Double -> JsonPrimitive(value)
        is Map<*, *> -> JsonObject(value.entries.associate { (k, v) -> k.toString() to toJsonElement(v) })
        is List<*> -> JsonArray(value.map { toJsonElement(it) })
        else -> JsonPrimitive(value.toString())
    }
}
